internal/pkg/pkg.go

```diff
@@ -0,0 +1,53 @@
+// Package pkg defines methods and heuristics for sanitizing package output.
+package pkg
+
+import (
+	"github.com/replit/upm/internal/api"
+)
+
+func makeLoweredHM(b api.LanguageBackend, names []string) map[api.PkgName]struct{} {
+	// Build a hashset. struct{}{} purportedly is of size 0, so this is as good as we get
+	set := make(map[api.PkgName]struct{})
+	for _, pkg := range names {
+		normal := b.NormalizePackageName(api.PkgName(pkg))
+		set[normal] = struct{}{}
+	}
+	return set
+}
+
+func SortPackages(b api.LanguageBackend, query string, ignoredPackages []string, packages []api.PkgInfo) []api.PkgInfo {
+
+	ignoredPackageSet := makeLoweredHM(b, ignoredPackages)
+
+	needle := b.NormalizePackageName(api.PkgName(query))
+	var exact *api.PkgInfo
+	prefixed := []api.PkgInfo{}
+	infixed := []api.PkgInfo{}
+	filtered := []api.PkgInfo{}
+
+	// If we have an exact package match, ensure it comes first.
+	for idx, pkg := range packages {
+		lower := b.NormalizePackageName(api.PkgName(pkg.Name))
+		if lower == needle {
+			exact = &packages[idx]
+		} else if _, isIgnored := ignoredPackageSet[lower]; isIgnored {
+			continue
+		} else if lower.HasPrefix(needle) {
+			prefixed = append(prefixed, pkg)
+		} else if lower.Contains(needle) {
+			infixed = append(infixed, pkg)
+		} else {
+			filtered = append(filtered, pkg)
+		}
+	}
+
+	curated := []api.PkgInfo{}
+	if exact != nil {
```

Comment:
How good is this scheme compared to the default? Have we analyze the quality of the results for some common queries? Are there situations where this scheme could demote some more desirable results?

Reply:
I think it depends on the backend, pypi surfaces particularly low quality results due to "relevance" being heavily weighted by the release date.

Maybe other package searches are better at this, to be sure, though in my experience `${ecosystem}-${extension}` is a common way to associate packages, with `${extension}-${ecosystem }` being less common but not impossible (thinking of a bridge between two different ecosystems, as an example). This is why I have both of them called out explicitly as higher ranking results.

By accumulating these sorted relevant packages and maintaining sub-ordering, my hope is that we have something of the best of both worlds, with exact name matches being on the top for relevance.

Reply:
Testing out `--language npm react`, most of the packages are still in the same results list, but stuff like `rxjs` is below the rest of the standard `react` libraries. A downside is that we don't search descriptions, so `styled-components` and `prop-types` are outside of the grouping.

Before
---
![image](https://github.com/replit/upm/assets/278900/14923842-0c2c-44a7-91b7-826954ab423f)

After
---
<img width="1420" alt="image" src="https://github.com/replit/upm/assets/278900/1b54b1a2-2b43-40c2-9473-9f35877efcf4">

Reply:
Since pypi is the one singled out as problematic, I'd prefer to only apply this to pypi.

Reply:
If it makes sense, I'd move the sorting logic into python's search function.